webapi/src/main/scala/org/knora/webapi/messages/admin/responder/projectsmessages/ProjectsMessagesADM.scala

```diff
@@ -505,130 +511,104 @@ case class ProjectADM(
-    )
+  def asShortnameOption: Option[String] =
+    self match {
+      case ShortnameIdentifier(value) => Some(value.value)
+      case _                          => None
+    }
 
-  /**
-   * Returns an optional value of the identifier.
-   */
-  def toIriOption: Option[IRI] =
-    maybeIri
+  def asUuidOption: Option[String] =
+    self match {
+      case UuidIdentifier(value) => Some(UuidIdentifier.makeProjectIri(value.value))
+      case _                     => None
+    }
+}
+
+object ProjectIdentifierADM {
 
   /**
-   * Tries to return the value as an SHORTNAME.
+   * Represents [[IriIdentifier]] identifier.
+   *
+   * @param value that constructs the identifier in the type of [[ProjectIri]] value object.
    */
-  def toShortname: String =
-    maybeShortname.getOrElse(
-      throw DataConversionException(s"Identifier $value is not of the required 'ProjectIdentifierType.SHORTNAME' type.")
-    )
+  final case class IriIdentifier(value: ProjectIri) extends ProjectIdentifierADM
+  object IriIdentifier {
+    def fromString(value: String): Validation[ValidationException, IriIdentifier] =
+      ProjectIri.make(value).map {
+        IriIdentifier(_)
+      }
+  }
 
   /**
-   * Returns an optional value of the identifier.
+   * Represents [[ShortcodeIdentifier]] identifier.
+   *
+   * @param value that constructs the identifier in the type of [[ShortCode]] value object.
    */
-  def toShortnameOption: Option[String] =
-    maybeShortname
+  final case class ShortcodeIdentifier(value: ShortCode) extends ProjectIdentifierADM
+  object ShortcodeIdentifier {
+    def fromString(value: String): Validation[ValidationException, ShortcodeIdentifier] =
+      ShortCode.make(value).map {
+        ShortcodeIdentifier(_)
+      }
+  }
 
   /**
-   * Tries to return the value as an SHORTCODE.
+   * Represents [[ShortnameIdentifier]] identifier.
+   *
+   * @param value that constructs the identifier in the type of [[ShortName]] value object.
    */
-  def toShortcode: String =
-    maybeShortcode.getOrElse(
-      throw DataConversionException(s"Identifier $value is not of the required 'ProjectIdentifierType.SHORTCODE' type.")
-    )
+  final case class ShortnameIdentifier(value: ShortName) extends ProjectIdentifierADM
+  object ShortnameIdentifier {
+    def fromString(value: String): Validation[ValidationException, ShortnameIdentifier] =
+      ShortName.make(value).map {
+        ShortnameIdentifier(_)
+      }
+  }
 
   /**
-   * Returns an optional value of the identifier.
+   * Represents [[UuidIdentifier]] identifier.
+   *
+   * @param value that constructs the identifier in the type of [[Base64Uuid]] value object.
    */
-  def toShortcodeOption: Option[String] =
-    maybeShortcode
+  final case class UuidIdentifier(value: Base64Uuid) extends ProjectIdentifierADM
+  // TODO-mpro: UuidIdentifier isn't fully used yet since the value object needs
```

Comment:
I don't understand this sentence (I think it is incomplete?)

Reply:
Not valid anymore, sorry forgot to remove it.